oak_crypto/src/noise_nk/rustcrypto.rs

```diff
@@ -0,0 +1,212 @@
+}
+
+pub fn aes_256_gcm_seal_in_place(
+    key: &[u8; 32],
+    nonce: &[u8; NONCE_LEN],
+    aad: &[u8],
+    plaintext: &mut Vec<u8>,
+) {
+    aes_gcm::Aes256Gcm::new_from_slice(key.as_slice())
+        .unwrap()
+        .encrypt_in_place(nonce.into(), aad, plaintext)
+        .unwrap();
+}
+
+pub fn aes_256_gcm_open_in_place(
+    key: &[u8; 32],
+    nonce: &[u8; NONCE_LEN],
+    aad: &[u8],
+    mut ciphertext: Vec<u8>,
+) -> Result<Vec<u8>, ()> {
+    aes_gcm::Aes256Gcm::new_from_slice(key.as_slice())
+        .unwrap()
+        .decrypt_in_place(nonce.into(), aad, &mut ciphertext)
+        .map_err(|_| ())?;
+    Ok(ciphertext)
+}
+
+pub fn sha256(input: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut ctx = sha2::Sha256::new();
+    ctx.update(input);
+    ctx.finalize().into()
+}
+
+/// Compute the SHA-256 hash of the concatenation of two inputs.
+pub fn sha256_two_part(input1: &[u8], input2: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut ctx = sha2::Sha256::new();
+    ctx.update(input1);
+    ctx.update(input2);
+    ctx.finalize().into()
+}
+
+pub struct P256Scalar {
+    v: p256::Scalar,
+}
+
+impl P256Scalar {
+    pub fn generate() -> P256Scalar {
+        let mut ret = [0u8; P256_SCALAR_LENGTH];
+        // Warning: not very random.
+        ret[0] = 1;
+        P256Scalar { v: p256::Scalar::from_repr(ret.into()).unwrap() }
+    }
+
+    pub fn compute_public_key(&self) -> [u8; P256_X962_LENGTH] {
+        p256::ProjectivePoint::mul_by_generator(&self.v)
+            .to_encoded_point(false)
+            .as_bytes()
+            .try_into()
+            .unwrap()
+    }
+
+    pub fn bytes(&self) -> [u8; P256_SCALAR_LENGTH] {
+        self.v.to_repr().as_slice().try_into().unwrap()
+    }
+}
+
+impl TryFrom<&[u8]> for P256Scalar {
+    type Error = ();
+
+    fn try_from(bytes: &[u8]) -> Result<Self, ()> {
+        let array: [u8; P256_SCALAR_LENGTH] = bytes.try_into().map_err(|_| ())?;
+        (&array).try_into()
+    }
+}
+
+impl TryFrom<&[u8; P256_SCALAR_LENGTH]> for P256Scalar {
+    type Error = ();
+
+    fn try_from(bytes: &[u8; P256_SCALAR_LENGTH]) -> Result<Self, ()> {
+        let scalar = p256::Scalar::from_repr((*bytes).into());
+        if !bool::from(scalar.is_some()) {
+            return Err(());
+        }
+        let scalar = scalar.unwrap();
+        if scalar.is_zero_vartime() {
+            return Err(());
+        }
+        Ok(P256Scalar { v: scalar })
+    }
+}
+
+pub fn p256_scalar_mult(
+    scalar: &P256Scalar,
+    point: &[u8; P256_X962_LENGTH],
+) -> Result<[u8; 32], ()> {
+    let point = p256::EncodedPoint::from_bytes(point).map_err(|_| ())?;
+    let affine_point = p256::AffinePoint::from_encoded_point(&point);
+    if !bool::from(affine_point.is_some()) {
+        // The peer's point is considered public input and so we don't need to work in
+        // constant time.
+        return Err(());
+    }
+    // unwrap: `is_some` checked just above.
```

Comment:
Indeed, therefore this should have been a pattern match instead.

Reply:
I don't think we can use pattern matching when (the enum's values are private)[https://docs.rs/subtle/latest/src/subtle/lib.rs.html#649].

Reply:
If you are ok with this not being constant time (which ideally would be clarified in a comment), I think you can transform it into a regular option and then do normal pattern matching. I haven't tried though